src/Infrastructure/VM/src/ESMCI_VM.C

```diff
@@ -1179,50 +1202,83 @@ void VM::shutdown(
+            if (matchTable_FObjects[i].size() > 0){
+              std::stringstream msg;
+              msg << "found corrupt garbage collection table";
+              ESMC_LogDefault.MsgFoundError(ESMC_RC_INTNRL_BAD, msg,
+                ESMC_CONTEXT, rc);
+              return;
+            }
+            if (garbageLog){
+              std::stringstream msg;
+              msg << "VM::finalize(): matchTable_FObjects[i] -   RELEASE";
+              ESMC_LogDefault.Write(msg, ESMC_LOGMSG_DEBUG);
+            }
+            // swap() trick with a temporary to free vector's memory
+            std::vector<FortranObject>().swap(matchTable_FObjects[i]);
           }
-#if 0
-// gjt: no take down garbage collection table for safety
-          if (matchTable_FObjects[i].size() > 0)
-            std::cout << "Failure in ESMF Automatic Garbage Collection line: "
-              << __LINE__ << std::endl;
-          // swap() trick with a temporary to free vector's memory
-          std::vector<FortranObject>().swap(matchTable_FObjects[i]);
-#endif
           // The following loop deletes deep C++ ESMF objects derived from
           // Base class. For deep Fortran classes it deletes the Base member.
           for (int k=matchTable_Objects[i].size()-1; k>=0; k--){
-#ifdef GARBAGE_COLLECTION_LOG_on
-            char msg[800];
-            const char *proxyString;
-            proxyString="actual";
-            if (matchTable_Objects[i][k]->ESMC_BaseGetProxyFlag()==ESMF_PROXYYES)
-              proxyString="proxy";
-            sprintf(msg, "ESMF Automatic Garbage Collection: c++base obj delete: "
-              "%20s %p - %6s - %7s - %7s : %04d : VM=%p : %10s",
-              matchTable_Objects[i][k]->ESMC_BaseGetClassName(),
-              matchTable_Objects[i][k], proxyString,
-              ESMC_StatusString(matchTable_Objects[i][k]->ESMC_BaseGetStatus()),
-              matchTable_Objects[i][k]->ESMC_BaseGetPersist() ?
-                "persist" : "noperst",
-              matchTable_Objects[i][k]->ESMC_BaseGetID(),
-              matchTable_Objects[i][k]->ESMC_BaseGetVM(),
-              matchTable_Objects[i][k]->ESMC_BaseGetName());
-            ESMC_LogDefault.Write(msg, ESMC_LOGMSG_DEBUG);
-#endif
-// gjt: no longer remove remnant from garbage collection for safety
-//            delete matchTable_Objects[i][k];  // delete ESMF object, incl. Base
-//            matchTable_Objects[i].pop_back();
+            bool mustDelete = false;  // safe mode is default
+            if (GlobalVM->garbageMode == garbageFull) mustDelete = true;
+            if (garbageLog){
+              char msg[800];
+              const char *proxyString;
+              proxyString="actual";
+              if (matchTable_Objects[i][k]->ESMC_BaseGetProxyFlag()==ESMF_PROXYYES)
+                proxyString="proxy";
+              const char *action;
+              action="KEEP";
+              if (mustDelete) action="DELETE";
```

Comment:
This is fine but slightly more efficient as an if...else

Reply:
Agree. Also seeing this again myself, I would have at least contracted lines 1245+1246 into `const char *action="KEEP";`. This came after moving some things around, so this is what it ended up no. Let's just keep it like this for now.